### DISCUSSION TITLE
ni_x_series_finite_sampling_input.py, acquire_frame: Effect of first clock pulse

### DISCUSSION THREAD

**Discussion by JohannesWesseler** [Ulm-IQO/qudi-iqo-modules#123] [Q&A] [answered]
Hi,
I am implementing a microwave generator in qudi and have a very short question regarding the ODMR functionality of qudi and specifically on the data acquisition during the frequency sweeps during ODMR.

The microwave has a trigger input that is connected to the physical_sample_clock_output. **My question:** Is the first clock pulse that is generated on that output channel supposed to either 
1) initialise the first microwave frequency, that is after the first clock pulse the microwave sits on the first frequency point
OR 
2) cause the microwave to jump from the 1. to the 2. microwave frequency already.

I could not really get that from the ni_x_series_finite_sampling_input.py or the odmr_logic.py files. 
It seems like the first clock pulse is directly generated when acquire_frame() is called, such that basically no time is spent on the first frequency point. So I would have assumed 1) is the answer, such that some time is also spent on the first frequency point. But I am not sure.
Getting this wrong would cause a frequency shift of the ODMR curve by 1 data point.

Thanks very much
Johannes

**Comment by Neverhorst** [comment, marked as answer]
Hi @JohannesWes ,

in all the microwave hardware modules i have seen, the call to `configure_scan()` or `start_scan()` will write the first frequency value 2 times into the jump list. This ensures there are as many frequency _transitions_ as trigger pulses per scan sweep. This is hardware implementation dependent and that's why it is handled in the hardware modules. If you think the docstrings in the interface need to reflect that, please feel free to propose a change.
E.g. for the R&S SMIQ devices you can see what I mean [here](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/6cd2ff89143e6d45f2c5b35c09c91d2d73ab9818/src/qudi/hardware/microwave/mw_source_smiq.py#L398).
